File: rust/ywta-link/src/presence.rs
//! YWTA Link v1 Peer Presence/Capability広告の型と検証を提供する。

use std::error::Error;
use std::fmt;

use serde::{Deserialize, Serialize};
use serde_json::Value;

/// Hello bodyで使用するversioned schema ID。
pub const PEER_HELLO_SCHEMA: &str = "ywta.peer.hello.v1";
/// Presence内の通常文字列に許可する最大Unicode文字数。
pub const PRESENCE_MAX_STRING_LENGTH: usize = 256;
/// Presenceが広告できるProtocol versionの最大数。
pub const PRESENCE_MAX_PROTOCOL_VERSIONS: usize = 16;
/// Presenceで広告できるProtocol versionの最大値。
pub const PRESENCE_MAX_PROTOCOL_VERSION: u16 = u16::MAX;
/// Presenceが広告できるCapabilityの最大数。
pub const PRESENCE_MAX_CAPABILITIES: usize = 128;
/// Capability IDに許可する最大Unicode文字数。
pub const PRESENCE_MAX_CAPABILITY_LENGTH: usize = 256;

/// Hello bodyとして広告するPeerの実装情報。
#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
#[serde(deny_unknown_fields)]
pub struct PeerPresence {
    pub peer_id: String,
    pub application: String,
    pub application_version: String,
    pub plugin_version: String,
    pub protocol_versions: Vec<u16>,
    pub capabilities: Vec<String>,
}

/// Peer Presence metadataがschemaに合わない理由。
#[derive(Clone, Debug, Eq, PartialEq)]
pub struct PresenceError {
    message: String,
}

impl PresenceError {
    /// 検証エラーを作成する。
    pub fn new(message: impl Into<String>) -> Self {
        Self {
            message: message.into(),
        }
    }
}

impl fmt::Display for PresenceError {
    fn fmt(&self, formatter: &mut fmt::Formatter<'_>) -> fmt::Result {
        formatter.write_str(&self.message)
    }
}

impl Error for PresenceError {}

impl PeerPresence {
    /// JSON objectのValueからPresenceを復元して検証する。
    pub fn from_value(value: &Value) -> Result<Self, PresenceError> {
        let presence: Self = serde_json::from_value(value.clone())
            .map_err(|error| PresenceError::new(format!("invalid peer hello body: {error}")))?;
        presence.validate()?;
        Ok(presence)
    }

    /// UTF-8 JSONからPresenceを復元して検証する。
    pub fn from_json(json: &[u8]) -> Result<Self, PresenceError> {
        let value: Value = serde_json::from_slice(json)
            .map_err(|error| PresenceError::new(format!("invalid peer hello JSON: {error}")))?;
        Self::from_value(&value)
    }

    /// PresenceをJSON Valueへ変換する。
    pub fn to_value(&self) -> Result<Value, PresenceError> {
        self.validate()?;
        serde_json::to_value(self)
            .map_err(|error| PresenceError::new(format!("cannot encode peer hello: {error}")))
    }

    /// Peer Presenceの各Fieldをfail closedで検証する。
    pub fn validate(&self) -> Result<(), PresenceError> {
        require_string(&self.peer_id, "peer_id", PRESENCE_MAX_STRING_LENGTH)?;
        require_string(&self.application, "application", PRESENCE_MAX_STRING_LENGTH)?;
        require_string(
            &self.application_version,
            "application_version",
            PRESENCE_MAX_STRING_LENGTH,
        )?;
        require_string(
            &self.plugin_version,
            "plugin_version",
            PRESENCE_MAX_STRING_LENGTH,
        )?;

        if self.protocol_versions.is_empty()
            || self.protocol_versions.len() > PRESENCE_MAX_PROTOCOL_VERSIONS
        {
            return Err(PresenceError::new(
                "protocol_versions has an invalid length",
            ));
        }
        if self
            .protocol_versions
            .windows(2)
            .any(|versions| versions[0] >= versions[1])
        {
            return Err(PresenceError::new(
                "protocol_versions must be sorted and unique",
            ));
        }
        if self.protocol_versions.contains(&0) {
            return Err(PresenceError::new(
                "protocol_versions must contain integers from 1 to 65535",
            ));
        }
        if !self.protocol_versions.contains(&1) {
            return Err(PresenceError::new(
                "protocol_versions must include version 1",
            ));
        }

        if self.capabilities.len() > PRESENCE_MAX_CAPABILITIES {
            return Err(PresenceError::new("capabilities has an invalid length"));
        }
        if self
            .capabilities
            .windows(2)
            .any(|capabilities| capabilities[0] >= capabilities[1])
        {
            return Err(PresenceError::new("capabilities must be sorted and unique"));
        }
        for capability in &self.capabilities {
            require_string(capability, "capability", PRESENCE_MAX_CAPABILITY_LENGTH)?;
            if !is_versioned_identifier(capability) {
                return Err(PresenceError::new(
                    "capabilities must contain non-empty versioned identifiers",
                ));
            }
        }
        Ok(())
    }
}

fn require_string(value: &str, field_name: &str, max_length: usize) -> Result<(), PresenceError> {
    if value.is_empty() || value.chars().count() > max_length {
        return Err(PresenceError::new(format!(
            "{field_name} must be a non-empty string of at most {max_length} characters"
        )));
    }
    Ok(())
}

fn is_versioned_identifier(value: &str) -> bool {
    let Some((prefix, version)) = value.rsplit_once(".v") else {
        return false;
    };
    if prefix.is_empty()
        || version.is_empty()
        || version.starts_with('0')
        || !version.bytes().all(|byte| byte.is_ascii_digit())
    {
        return false;
    }
    prefix.split('.').all(|segment| {
        let mut characters = segment.bytes();
        matches!(characters.next(), Some(first) if first.is_ascii_alphabetic())
            && characters.all(|character| {
                character.is_ascii_alphanumeric() || matches!(character, b'_' | b'-')
            })
    })
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parses_shared_golden_fixture() {
        let presence = PeerPresence::from_json(include_bytes!(
            "../../../tests/link/fixtures/peer_hello_v1.json"
        ))
        .expect("golden fixture must decode");

        assert_eq!(presence.peer_id, "blender:peer-001");
        assert_eq!(presence.protocol_versions, vec![1]);
        assert_eq!(
            presence.capabilities,
            vec![
                "camera.apply.v1".to_owned(),
                "camera.read.v1".to_owned(),
                "transform.read.v1".to_owned(),
            ]
        );
    }

    #[test]
    fn rejects_unknown_fields_and_unsorted_metadata() {
        let unknown = serde_json::json!({
            "peer_id": "blender:peer-001",
            "application": "Blender",
            "application_version": "4.5.0",
            "plugin_version": "0.1.0",
            "protocol_versions": [1],
            "capabilities": ["camera.read.v1"],
            "unexpected": true,
        });
        assert!(PeerPresence::from_value(&unknown).is_err());

        let unsorted = serde_json::json!({
            "peer_id": "blender:peer-001",
            "application": "Blender",
            "application_version": "4.5.0",
            "plugin_version": "0.1.0",
            "protocol_versions": [2, 1],
            "capabilities": ["camera.read.v1"],
        });
        assert!(PeerPresence::from_value(&unsorted).is_err());
    }

    #[test]
    fn accepts_protocol_upper_bound_and_empty_capabilities() {
        let value = serde_json::json!({
            "peer_id": "blender:peer-001",
            "application": "Blender",
            "application_version": "4.5.0",
            "plugin_version": "0.1.0",
            "protocol_versions": [1, PRESENCE_MAX_PROTOCOL_VERSION],
            "capabilities": [],
        });
        let presence = PeerPresence::from_value(&value).expect("boundary metadata must decode");

        assert_eq!(presence.protocol_versions, vec![1, 65535]);
        assert!(presence.capabilities.is_empty());
    }

    #[test]
    fn rejects_zero_overflow_and_mixed_protocol_or_capability_values() {
        let base = serde_json::json!({
            "peer_id": "blender:peer-001",
            "application": "Blender",
            "application_version": "4.5.0",
            "plugin_version": "0.1.0",
            "protocol_versions": [1],
            "capabilities": [],
        });
        for versions in [
            serde_json::json!([0, 1]),
            serde_json::json!([1, 65536]),
            serde_json::json!([true]),
        ] {
            let mut value = base.clone();
            value["protocol_versions"] = versions;
            assert!(PeerPresence::from_value(&value).is_err());
        }
        for capabilities in [
            serde_json::json!(["camera.read.v1", 1]),
            serde_json::json!([[]]),
        ] {
            let mut value = base.clone();
            value["capabilities"] = capabilities;
            assert!(PeerPresence::from_value(&value).is_err());
        }
    }
}

File: ywta_link/presence.py
"""YWTA Link v1 Peer Presence/Capability広告の型と検証。"""

from __future__ import annotations

import json
import re
from dataclasses import dataclass
from typing import Any, Mapping

from .errors import ValidationError

PEER_HELLO_SCHEMA = "ywta.peer.hello.v1"
PRESENCE_MAX_STRING_LENGTH = 256
PRESENCE_MAX_PROTOCOL_VERSIONS = 16
PRESENCE_MAX_PROTOCOL_VERSION = 65535
PRESENCE_MAX_CAPABILITIES = 128
PRESENCE_MAX_CAPABILITY_LENGTH = 256

_PRESENCE_FIELDS = frozenset(
    {
        "peer_id",
        "application",
        "application_version",
        "plugin_version",
        "protocol_versions",
        "capabilities",
    }
)
_VERSIONED_ID = re.compile(r"^[A-Za-z][A-Za-z0-9_-]*(?:\.[A-Za-z][A-Za-z0-9_-]*)*\.v[1-9][0-9]*$")


class PresenceValidationError(ValidationError):
    """Peer Presence metadataの検証失敗。"""


@dataclass(frozen=True)
class PeerPresence:
    """Helloに含める不変なPeer ID、実装情報、Capability広告。"""

    peer_id: str
    application: str
    application_version: str
    plugin_version: str
    protocol_versions: tuple[int, ...]
    capabilities: tuple[str, ...]

    def __post_init__(self) -> None:
        """構築時にschemaの全制約を検証する。"""

        self.validate()

    def validate(self) -> "PeerPresence":
        """広告を送信可能な値として再検証する。"""

        _required_string(self.peer_id, "peer_id")
        _required_string(self.application, "application")
        _required_string(self.application_version, "application_version")
        _required_string(self.plugin_version, "plugin_version")
        if not isinstance(self.protocol_versions, tuple):
            raise PresenceValidationError("protocol_versions must be an immutable tuple")
        if not 0 < len(self.protocol_versions) <= PRESENCE_MAX_PROTOCOL_VERSIONS:
            raise PresenceValidationError("protocol_versions has an invalid length")
        if any(
            isinstance(version, bool) or not isinstance(version, int) or not 1 <= version <= PRESENCE_MAX_PROTOCOL_VERSION
            for version in self.protocol_versions
        ):
            raise PresenceValidationError(f"protocol_versions must contain integers from 1 to {PRESENCE_MAX_PROTOCOL_VERSION}")
        if self.protocol_versions != tuple(sorted(set(self.protocol_versions))):
            raise PresenceValidationError("protocol_versions must be sorted and unique")
        if 1 not in self.protocol_versions:
            raise PresenceValidationError("protocol_versions must include version 1")
        if not isinstance(self.capabilities, tuple):
            raise PresenceValidationError("capabilities must be an immutable tuple")
        if len(self.capabilities) > PRESENCE_MAX_CAPABILITIES:
            raise PresenceValidationError("capabilities has an invalid length")
        for capability in self.capabilities:
            _required_string(
                capability,
                "capability",
                max_length=PRESENCE_MAX_CAPABILITY_LENGTH,
            )
            if not _VERSIONED_ID.fullmatch(capability):
                raise PresenceValidationError("capabilities must contain non-empty versioned identifiers")
        if self.capabilities != tuple(sorted(set(self.capabilities))):
            raise PresenceValidationError("capabilities must be sorted and unique")
        return self

    @classmethod
    def from_dict(cls, value: Mapping[str, Any]) -> "PeerPresence":
        """JSON objectからPeer Presenceを復元し、未知Fieldを拒否する。"""

        if not isinstance(value, Mapping):
            raise PresenceValidationError("peer hello body must be a JSON object")
        unknown = set(value) - _PRESENCE_FIELDS
        missing = _PRESENCE_FIELDS - set(value)
        if unknown:
            raise PresenceValidationError(f"unknown peer hello fields: {sorted(unknown)}")
        if missing:
            raise PresenceValidationError(f"missing peer hello fields: {sorted(missing)}")
        protocol_versions = value["protocol_versions"]
        capabilities = value["capabilities"]
        if not isinstance(protocol_versions, list):
            raise PresenceValidationError("protocol_versions must be an array")
        if not isinstance(capabilities, list):
            raise PresenceValidationError("capabilities must be an array")
        if len(protocol_versions) > PRESENCE_MAX_PROTOCOL_VERSIONS:
            raise PresenceValidationError("protocol_versions has an invalid length")
        if len(capabilities) > PRESENCE_MAX_CAPABILITIES:
            raise PresenceValidationError("capabilities has an invalid length")
        return cls(
            peer_id=value["peer_id"],
            application=value["application"],
            application_version=value["application_version"],
            plugin_version=value["plugin_version"],
            protocol_versions=tuple(protocol_versions),
            capabilities=tuple(capabilities),
        )

    @classmethod
    def decode(cls, payload: str | bytes | bytearray) -> "PeerPresence":
        """UTF-8 JSONからPeer Presenceを復元する。"""

        try:
            value = json.loads(payload)
        except (TypeError, ValueError, UnicodeDecodeError) as exc:
            raise PresenceValidationError(f"invalid peer hello JSON: {exc}") from exc
        return cls.from_dict(value)

    def to_dict(self) -> dict[str, Any]:
        """Peer PresenceをJSON objectへ変換する。"""

        self.validate()
        return {
            "peer_id": self.peer_id,
            "application": self.application,
            "application_version": self.application_version,
            "plugin_version": self.plugin_version,
            "protocol_versions": list(self.protocol_versions),
            "capabilities": list(self.capabilities),
        }

    def encode(self) -> str:
        """決定的なUTF-8 JSON文字列へ変換する。"""

        try:
            return json.dumps(
                self.to_dict(),
                ensure_ascii=False,
                separators=(",", ":"),
                sort_keys=True,
                allow_nan=False,
            )
        except (TypeError, ValueError) as exc:
            raise PresenceValidationError(f"cannot encode peer hello: {exc}") from exc


def _required_string(
    value: object,
    field_name: str,
    *,
    max_length: int = PRESENCE_MAX_STRING_LENGTH,
) -> None:
    """Presence内の文字列Fieldを検証する。"""

    if not isinstance(value, str) or not value or len(value) > max_length:
        raise PresenceValidationError(f"{field_name} must be a non-empty string of at most {max_length} characters")
    try:
        value.encode("utf-8")
    except UnicodeEncodeError as exc:
        raise PresenceValidationError(f"{field_name} must be valid UTF-8") from exc
